### PR TITLE
[WIP] Store and Model Refactor

### PR DESCRIPTION
**Pull request by hydrabolt**
## Current Issue
**TL;DR bad design patterns haven't died yet**
1) `ClientDataManager` is a horrible design decision of mine - it shouldn't _still_ exist. It acts as a controller for collections all over the place (note that they're collections, we don't even formally refer to them as stores!)
2) Many of the data structures obviously have things in common (many are ID-based, have a similar instantiation process and some structures are able to be updated by Discord update events.) Despite this, the data structures do not extend from a base class. This means updating structures requires a lot of repeated code, as does instantiation.

## Improving Data Structures
This PR creates a new `Base` class for classes that represent data models used in the Discord API. This class automatically sets an immutable reference to the client (used to be repeated in many of our structures). The current list of classes extending the `Base` class: `Channel`, `Emoji`, `Guild`, `GuildMember`, `Invite`, `Message`, `OAuth2Application`, `PartialGuild`, `PartialGuildChannel`, `Role`, `User`, and `UserProfile`.

The `Base` class also fixes some current issues, such as repeated code when updating instances. It features the following methods:

1) `_clone()` returns a shallow clone of the instance.
2) `_patch(data)` is introduced to attempt to bring order to the whole `setup()` madness. This new method is private (developers needn't mess with it) and is slightly better-named.
3) `_update(data)` is a combination of both methods. It patches the current instance, but returns a cloned  version prior to the patch. This can replace similar updating logic in the rest of the project.

Classes extending Base should set reusable information in the `_patch` method and then call `_patch` in the constructor with the data.

## Implementing Data Stores
Instead of creating a giant single data store (_cough_ #1503 and #1628), I decided on creating an extendable `DataStore` class (that extends `Collection` for backwards-compatibility.) This class stores a reference to the client and implements `create()` and `remove()` stubs, which take data straight from the Discord API and create and store the appropriate data models.

Rather than creating `DataStore` classes directly, extra classes are created, e.g. `UserStore`, `GuildStore`, `ChannelStore` which implement their own logic for the creation and removal of data structures (as they need to do different things.) These classes can then replace the collections found in `Client` and are still used in the same fashion for the developer.

This also removes the need for `ClientDataManager`. Instead of `client.dataManager.newUser(data)`, we can now use `client.users.create(data)` which is much cleaner and less convoluted.

## Summary
1) Most structures extend a Base class featuring extra magic
2) Updating these structures is easier
3) `ClientDataManger` has finally been slain and replaced with data stores.

~~Surprisingly, this is not a breaking change.~~ This is a breaking change:

1) `client.fetchUser` -> `client.users.fetch`

### CODE REVIEW

**Comment by devsnek** [comment]
so like... fuck #1503?

**Comment by hydrabolt** [comment]
yes @devsnek

**Comment by bdistin** [comment]
client.fetchUser -> client.users.fetch is 👍

File: src/structures/Base.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * The client that instantiated this
+     * @name Base#client
+     * @type {Client}
+     * @readonly
+     */
+    Object.defineProperty(this, 'client', {
+      value: client,
+      enumerable: false,
+      writable: false,
```
Comment (appellation) [inline src/structures/Base.js]: more extraneous `enumerable`/`writable` definition?

**Comment by Drahcirius** [comment]
There will be a message store added as well, right?

**Comment by hydrabolt** [comment]
Yep @Drahcirius

File: src/stores/DataStore.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * @extends {Collection}
+ */
+class DataStore extends Collection {
+  constructor(client, iterable) {
+    super(iterable);
+
+    Object.defineProperty(this, 'client', {
+      value: client,
+      enumerable: false,
+      writable: false,
```
Comment (hydrabolt) [inline src/stores/DataStore.js]: fixed

File: src/stores/DataStore.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+const Collection = require('../util/Collection');
+
+/**
+ * A data store what else????
```
Comment (hydrabolt) [inline src/stores/DataStore.js]: fixed

File: src/structures/Base.js (hunk 1 above)
Comment (hydrabolt) [inline src/structures/Base.js]: fixed

**Comment by hydrabolt** [comment]
A quick follow-up on this, all that's left to do (I think):

- Role data store
- Don't return unnecessary data in actions

File: src/structures/GuildMember.js [hunk 4]
```diff
@@ -2,22 +2,17 @@ const TextBasedChannel = require('./interfaces/TextBasedChannel');
...
 const Permissions = require('../util/Permissions');
 const Collection = require('../util/Collection');
+const Base = require('./Base');
 const { Presence } = require('./Presence');
 const { Error, TypeError } = require('../errors');
 
 /**
  * Represents a member of a guild on Discord.
  * @implements {TextBasedChannel}
  */
```
Comment (SpaceEEC) [inline src/structures/GuildMember.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/Invite.js [hunk 5]
```diff
@@ -1,26 +1,19 @@
 const Constants = require('../util/Constants');
+const Base = require('./Base');
 
 /**
  * Represents an invitation to a guild channel.
  * <warn>The only guaranteed properties are `code`, `guild` and `channel`. Other properties can be missing.</warn>
  */
```
Comment (SpaceEEC) [inline src/structures/Invite.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/Message.js [hunk 6]
```diff
@@ -6,34 +6,30 @@ const ReactionCollector = require('./ReactionCollector');
...
+const ReactionStore = require('../stores/ReactionStore');
 const Constants = require('../util/Constants');
 const Permissions = require('../util/Permissions');
+const Base = require('./Base');
 const { Error, TypeError } = require('../errors');
 let GuildMember;
 
 /**
  * Represents a message on Discord.
  */
```
Comment (SpaceEEC) [inline src/structures/Message.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/Message.js [hunk 7]
```diff
@@ -118,7 +114,7 @@ class Message {
      * A collection of reactions to this message, mapped by the reaction ID
      * @type {Collection<Snowflake, MessageReaction>}
```
Comment (SpaceEEC) [inline src/structures/Message.js]: Should be documented as a `ReactionStore` and not as a `Collection`.

File: src/structures/Role.js [hunk 8]
```diff
@@ -1,30 +1,26 @@
 const Snowflake = require('../util/Snowflake');
 const Permissions = require('../util/Permissions');
 const Util = require('../util/Util');
+const Base = require('./Base');
 
 /**
  * Represents a role on Discord.
  */
```
Comment (SpaceEEC) [inline src/structures/Role.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/User.js [hunk 9]
```diff
@@ -3,26 +3,21 @@ const Constants = require('../util/Constants');
 const { Presence } = require('./Presence');
 const UserProfile = require('./UserProfile');
 const Snowflake = require('../util/Snowflake');
+const Base = require('./Base');
 const { Error } = require('../errors');
 
 /**
  * Represents a user on Discord.
  * @implements {TextBasedChannel}
  */
```
Comment (SpaceEEC) [inline src/structures/User.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/UserProfile.js [hunk 10]
```diff
@@ -1,27 +1,22 @@
 const Collection = require('../util/Collection');
 const { UserFlags } = require('../util/Constants');
 const UserConnection = require('./UserConnection');
+const Base = require('./Base');
 
 /**
  * Represents a user's profile on Discord.
  */
```
Comment (SpaceEEC) [inline src/structures/UserProfile.js]: `@extends {Base}`, otherwise the `client` property is not visible in the docs.

File: src/structures/interfaces/TextBasedChannel.js [hunk 11]
```diff
@@ -17,7 +18,7 @@ class TextBasedChannel {
      * A collection containing the messages sent to this channel
      * @type {Collection<Snowflake, Message>}
```
Comment (SpaceEEC) [inline src/structures/interfaces/TextBasedChannel.js]: Should be documented as a `MessageStore` and not as a `Collection`.

File: src/stores/DataStore.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
+const Collection = require('../util/Collection');
+
+/**
+ * A data store
+ * @class DataStore
```
Comment (SpaceEEC) [inline src/stores/DataStore.js]: Is `@class` necessary?
We have that nowhere else in the documentation.

File: src/structures/interfaces/TextBasedChannel.js [hunk 13]
```diff
@@ -148,7 +149,6 @@ class TextBasedChannel {
    *   .catch(console.error);
    */
   fetchMessage(messageID) {
```
Comment (SpaceEEC) [inline src/structures/interfaces/TextBasedChannel.js]: If this method is intended to be replaced by `MessageStore#fetch`, remove.

File: src/structures/interfaces/TextBasedChannel.js [hunk 14]
```diff
@@ -186,15 +182,10 @@ class TextBasedChannel {
    *   .catch(console.error);
    */
```
Comment (SpaceEEC) [inline src/structures/interfaces/TextBasedChannel.js]: If this method is intended to be replaced by `MessageStore#fetch`, remove.

File: src/structures/interfaces/TextBasedChannel.js [hunk 15]
```diff
@@ -204,14 +195,9 @@ class TextBasedChannel {
    * @returns {Promise<Collection<Snowflake, Message>>}
    */
   fetchPinnedMessages() {
```
Comment (SpaceEEC) [inline src/structures/interfaces/TextBasedChannel.js]: If `fetchMessage` and `fetchMessages` are meant to be moved to `MessageStore` this method could also be moved there.

File: src/structures/User.js [hunk 16]
```diff
@@ -3,26 +3,21 @@ const Constants = require('../util/Constants');
...
-     */
-    Object.defineProperty(this, 'client', { value: client });
-
-    if (data) this.setup(data);
+    super(client);
+    this._patch(data);
   }
 
-  setup(data) {
+  _patch(data) {
```
Comment (SpaceEEC) [inline src/structures/User.js]: PresenceUpdate packets (maybe others too) are calling this method with partial data, which somewhat breaks the user cache, for reference:
```js
User {
id: '173547401905176585',
username: undefined,
discriminator: undefined,
avatar: undefined,
bot: false,
lastMessageID: null,
lastMessage: null }
```
For example using the "old" `patch` method for patching with potential partial data and moving the initial property assigning to the constructor (or even to a/the setup method) would solve that issue.

File: src/stores/MessageStore.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    const message = new Message(this.client.channels.get(data.channel_id), data, this.client);
+
+    if (cache) this.set(message.id, message);
+    return message;
+  }
+
+  set(key, value) {
+    super.set(key, value);
+    const maxSize = this.client.options.messageCacheMaxSize;
+    if (maxSize === 0) this.delete(key);
```
Comment (bdistin) [inline src/stores/MessageStore.js]: Shouldn't this be checked before setting, and just return if maxSize is 0?

File: src/structures/User.js (hunk 16 above)
Comment (hydrabolt) [inline src/structures/User.js]: Fixed

File: src/stores/DataStore.js (hunk 12 above)
Comment (hydrabolt) [inline src/stores/DataStore.js]: Fixed

File: src/stores/ChannelStore.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
+const DataStore = require('./DataStore');
+const DMChannel = require('../structures/DMChannel');
+const GroupDMChannel = require('../structures/GroupDMChannel');
+const Constants = require('../util/Constants');
+
+class ChannelStore extends DataStore {
+  create(data, guild, cache = true) {
+    super.create();
```
Comment (bdistin) [inline src/stores/ChannelStore.js]: DataStore.create returns undefined (and really does nothing). If you were abstracting some of or all of the creation/removal then this would be great. But as it stands, this is unnecessary.
I will refer to the rest as DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/ChannelStore.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+        break;
+    }
+
+    if (cache) this.set(channel.id, channel);
+
+    return channel;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/ChannelStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/EmojiStore.js [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+const Emoji = require('../structures/Emoji');
+
+class EmojiStore extends DataStore {
+  constructor(guild, iterable) {
+    super(guild.client, iterable);
+    this.guild = guild;
+  }
+
+  create(data) {
+    super.create();
```
Comment (bdistin) [inline src/stores/EmojiStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/EmojiStore.js [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+    if (existing) return existing;
+
+    const emoji = new Emoji(guild, data);
+    guild.emojis.set(emoji.id, emoji);
+
+    return emoji;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/EmojiStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/GuildChannelStore.js [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+const Constants = require('../util/Constants');
+
+class GuildChannelStore extends DataStore {
+  constructor(guild, iterable) {
+    super(guild.client, iterable);
+    this.guild = guild;
+  }
+
+  create(data, cache = true) {
+    super.create();
```
Comment (bdistin) [inline src/stores/GuildChannelStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/GuildStore.js [hunk 23]
```diff
@@ -0,0 +1,23 @@
+const DataStore = require('./DataStore');
+const Guild = require('../structures/Guild');
+
+class GuildStore extends DataStore {
+  create(data) {
+    super.create();
```
Comment (bdistin) [inline src/stores/GuildStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/GuildStore.js [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    if (existing) return existing;
+
+    const guild = new Guild(this.client, data);
+    this.set(guild.id, guild);
+
+    return guild;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/GuildStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/MessageStore.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+class MessageStore extends DataStore {
+  constructor(channel, iterable) {
+    super(channel.client, iterable);
+    this.channel = channel;
+    Message = require('../structures/Message');
+  }
+
+  create(data, cache = true) {
+    super.create();
```
Comment (bdistin) [inline src/stores/MessageStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/MessageStore.js [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+
+  set(key, value) {
+    super.set(key, value);
+    const maxSize = this.client.options.messageCacheMaxSize;
+    if (maxSize === 0) this.delete(key);
+    if (this.size >= maxSize && maxSize > 0) this.delete(this.firstKey());
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/MessageStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/ReactionStore.js [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+const MessageReaction = require('../structures/MessageReaction');
+
+class ReactionStore extends DataStore {
+  constructor(message, iterable) {
+    super(message.client, iterable);
+    this.message = message;
+  }
+
+  create(data) {
+    super.create();
```
Comment (bdistin) [inline src/stores/ReactionStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/ReactionStore.js [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+    if (existing) return existing;
+
+    const reaction = new MessageReaction(this.message, data.emoji, data.count, data.me);
+    this.set(emojiID, reaction);
+
+    return reaction;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/ReactionStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/UserStore.js [hunk 29]
```diff
@@ -0,0 +1,38 @@
+const DataStore = require('./DataStore');
+const User = require('../structures/User');
+
+class UserStore extends DataStore {
+  create(data) {
+    super.create();
```
Comment (bdistin) [inline src/stores/UserStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/UserStore.js [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+    const existing = this.get(data.id);
+    if (existing) return existing;
+
+    const user = new User(this.client, data);
+    this.set(user.id, user);
+    return user;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/UserStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/stores/GuildChannelStore.js [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+
+    const ChannelModel = data.type === Constants.ChannelTypes.TEXT ? TextChannel : VoiceChannel;
+    const channel = new ChannelModel(this.guild, data);
+    if (cache) this.set(channel.id, channel);
+
+    return channel;
+  }
+
+  remove(id) {
+    super.remove();
```
Comment (bdistin) [inline src/stores/GuildChannelStore.js]: DATASTORE_CREATE_REMOVE_SUPER

File: src/structures/interfaces/TextBasedChannel.js (hunk 15 above)
Comment (hydrabolt) [inline src/structures/interfaces/TextBasedChannel.js]: Fixed

File: src/structures/interfaces/TextBasedChannel.js (hunk 13 above)
Comment (hydrabolt) [inline src/structures/interfaces/TextBasedChannel.js]: Fixed

File: src/structures/interfaces/TextBasedChannel.js (hunk 14 above)
Comment (hydrabolt) [inline src/structures/interfaces/TextBasedChannel.js]: Fixed

File: src/structures/User.js (hunk 9 above)
Comment (hydrabolt) [inline src/structures/User.js]: Fixed

File: src/structures/UserProfile.js (hunk 10 above)
Comment (hydrabolt) [inline src/structures/UserProfile.js]: Fixed

File: src/structures/Role.js (hunk 8 above)
Comment (hydrabolt) [inline src/structures/Role.js]: Fixed

File: src/structures/Message.js (hunk 6 above)
Comment (hydrabolt) [inline src/structures/Message.js]: Fixed

File: src/structures/Invite.js (hunk 5 above)
Comment (hydrabolt) [inline src/structures/Invite.js]: Fixed

File: src/structures/GuildMember.js (hunk 4 above)
Comment (hydrabolt) [inline src/structures/GuildMember.js]: Fixed

File: src/structures/Guild.js [hunk 32]
```diff
@@ -15,22 +15,19 @@ const Util = require('../util/Util');
...
+const EmojiStore = require('../stores/EmojiStore');
+const GuildChannelStore = require('../stores/GuildChannelStore');
+const Base = require('./Base');
 const { Error, TypeError } = require('../errors');
 
 /**
  * Represents a guild (or a server) on Discord.
  * <info>It's recommended to see if a guild is available before performing operations or reading data from it. You can
  * check this with `guild.available`.</info>
  */
```
Comment (hydrabolt) [inline src/structures/Guild.js]: Fixed

File: src/structures/Emoji.js [hunk 33]
```diff
@@ -1,30 +1,26 @@
 const Constants = require('../util/Constants');
 const Collection = require('../util/Collection');
 const Snowflake = require('../util/Snowflake');
+const Base = require('./Base');
 
 /**
  * Represents a custom emoji.
  */
```
Comment (hydrabolt) [inline src/structures/Emoji.js]: Fixed

File: src/structures/ClientApplication.js [hunk 34]
```diff
@@ -1,23 +1,18 @@
 const Snowflake = require('../util/Snowflake');
 const Constants = require('../util/Constants');
+const Base = require('./Base');
 
 /**
  * Represents a Client OAuth2 Application.
  */
```
Comment (hydrabolt) [inline src/structures/ClientApplication.js]: Fixed

File: src/structures/Channel.js [hunk 35]
```diff
@@ -1,18 +1,13 @@
 const Snowflake = require('../util/Snowflake');
+const Base = require('./Base');
 const Constants = require('../util/Constants');
 
 /**
  * Represents any channel on Discord.
  */
```
Comment (hydrabolt) [inline src/structures/Channel.js]: Fixed

File: src/stores/UserStore.js [hunk 36]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  /**
+   * Obtains a user from Discord, or the user cache if it's already available.
+   * <warn>This is only available when using a bot account.</warn>
+   * @param {Snowflake} id ID of the user
+   * @param {boolean} [cache=true] Whether to cache the new user object if it isn't already
+   * @returns {Promise<User>}
+   */
+  fetch(id, cache = true) {
```
Comment (hydrabolt) [inline src/stores/UserStore.js]: Fixed

File: src/stores/MessageStore.js (hunk 17 above)
Comment (hydrabolt) [inline src/stores/MessageStore.js]: Can't access `this` until super is called

File: src/structures/GuildMember.js [hunk 37]
```diff
@@ -99,13 +99,13 @@ class GuildMember extends Base {
...
   /**
-   * The voice session ID of this member, if any
+   * The voice session ID of this member (if any)
    * @type {?Snowflake}
    */
   get voiceSessionID() { return this.voiceState.session_id; }
 
   /**
-   * The voice channel ID of this member, if any
+   * The voice channel ID of this member, (if any)
```
Comment (KingDGrizzle) [inline src/structures/GuildMember.js]: That comma shouldn't be there
